sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -637,3 +638,207 @@ def test_categories(density, drop):
+    int_category = pd.api.types.CategoricalDtype(
+         categories=[3, 1, 2], ordered=True)
+
+    X_df['col_str'] = X_df['col_str'].astype(str_category)
+    X_df['col_int'] = X_df['col_int'].astype(int_category)
+
+    ohe = OneHotEncoder(categories='dtypes', sparse=is_sparse, drop=drop)
+
+    expected_trans = np.array([
+        [0, 1, 1, 0, 0],
+        [1, 0, 0, 0, 1],
+        [1, 0, 0, 1, 0],
+        [0, 1, 0, 0, 1]], dtype=np.float64)
+
+    if drop == 'first':
+        expected_trans = expected_trans[:, [1, 3, 4]]
+
+    X_trans = ohe.fit_transform(X_df)
+    if is_sparse:
+        X_trans = X_trans.toarray()
+
+    assert_allclose(X_trans, expected_trans)
+
+    X_inverse = ohe.inverse_transform(expected_trans)
+
+    assert_array_equal(X_inverse, X_df.values)
+
+
+@pytest.mark.parametrize('encoder', [
+    OneHotEncoder(categories="dtypes"),
+    OrdinalEncoder(categories="dtypes")])
+def test_encoder_pd_error_mismatch_dtype(encoder):
+    pd = pytest.importorskip('pandas')
+    msg = "X.dtypes must match the dtypes used when fitting"
+
+    X_df_orig = pd.DataFrame({
+        'col_str': ['a', 'b', 'b', 'a'],
+        'col_int': [3, 2, 1, 2]}, columns=['col_str', 'col_int'])
+
+    enc_no_categories = clone(encoder).fit(X_df_orig)
+
+    X_df0 = X_df_orig.copy()
+    X_df0['col_int'] = X_df0['col_int'].astype('category')
+
+    # X_df0 has categories while the trained dataframe does not
+    with pytest.raises(ValueError, match=msg):
+        enc_no_categories.transform(X_df0)
+
+    str_category = pd.api.types.CategoricalDtype(
+         categories=['b', 'a'], ordered=True)
+    X_df1 = X_df_orig.copy()
+    X_df1['col_str'] = X_df1['col_str'].astype(str_category)
+    X_df1['col_int'] = X_df1['col_int'].astype('category')
+
+    # X_df1 has categories while the trained dataframe does not
+    with pytest.raises(ValueError, match=msg):
+        enc_no_categories.transform(X_df1)
+
+    # Train encoder with categoricals
+    enc = clone(encoder).fit(X_df1)
+
+    # col_str dtype not ordered correctly
+    X_df2 = X_df_orig.copy()
+    str_category_lex_ordered = pd.api.types.CategoricalDtype(
+         categories=['a', 'b'], ordered=True)
+    X_df2['col_str'] = X_df2['col_str'].astype(str_category_lex_ordered)
+    X_df2['col_int'] = X_df2['col_int'].astype('category')
+
+    with pytest.raises(ValueError, match=msg):
+        enc.transform(X_df2)
+
+    # col_int not a categorical dtype
+    X_df3 = X_df_orig.copy()
+    X_df3['col_int'] = X_df3['col_int'].astype(int)
+
+    with pytest.raises(ValueError, match=msg):
+        enc.transform(X_df3)
+
+    # number of features is not correct
+    X_df4 = pd.DataFrame({
+        'col_str': ['a', 'b', 'b', 'a']})
+
+    with pytest.raises(ValueError, match=msg):
+        enc.transform(X_df4)
+
+
+@pytest.mark.parametrize("is_sparse", [True, False])
+@pytest.mark.parametrize("drop", ["first", None])
+def test_one_hot_encoder_pd_categories_mixed(is_sparse, drop):
```

Comment:
You could also add one non-category column to the first test, to avoid some redundancy. 
The dataframe is handled column by column anyway (it's good to have a test for this of course, but just think that an additional test is not really necessary, a dataframe with only categorical columns is not a special case)